Move Reserve and MiqExpression to lib folder

lpichler opened the pull request:
Purpose or Intent
-----------------
Some models need to be skipped from good_migrations detection because there are autoloaded (typically) somewhere deeply in code and migration stubs are not taken into account.

So I am moving **MiqExpression** and **Reserve** to /lib folder.
With MiqExpression there were moved also these files:
```
app/models/miq_expression/tag.rb 
app/models/miq_expression/relative_datetime.rb
app/models/miq_expression/field.rb 
```

also specs for this files were moved:
### MiqExpression
`spec/models/miq_expression_spec.rb` to `spec/lib/miq_expression_spec.rb`
`spec/models/miq_expression/field_spec.rb` to `spec/lib/miq_expression/field_spec.rb`
`spec/models/miq_expression/relative_datetime_spec.rb` to `spec/lib/miq_expression/relative_datetime_spec.rb`
no other related files exist.
### Reserve
there is no reserve_spec so I moved `reserved_mixin_spec.rb` to `spec/lib/reserved_mixin_spec.rb` as we moved reserved_mixin in https://github.com/ManageIQ/manageiq/pull/7348

----------------------------------

There are 2 cases in these PR:

1) Loaded by Yaml Loader 
Mostly MiqExpression model is autoloaded by yaml loader here:
https://github.com/ManageIQ/manageiq/blob/master/config/initializers/yaml_autoloader.rb#L6
and it happens always from app/models folder

In Yaml we have
```
--- !ruby/object:MiqExpression
exp:
  IS NULL:
    field: Vm.hardware-bios
    value: 
```
and yaml loader create object of MiqExpression

It seems that  there can be any model  but
It was detected only in migration 
https://github.com/ManageIQ/manageiq/blob/master/db/migrate/20150625220141_fix_serialized_reports_for_rails_four.rb
where data migration are manipulating with MiqReportResult#report where serialized report and there are only MiqExpression in `!ruby/object:` statements in report definition.
Correct me if I am wrong.

2)  Reserve Model
when we are using ReservedMixin in migrations (https://github.com/ManageIQ/manageiq/pull/9827) we using Reserved table as relation.
for example:
https://github.com/ManageIQ/manageiq/pull/9827/commits/37399fdf138a2486bc65d2c4c2c9e3420925a643#diff-dd59f104a6dc6bfd871a7fa1e2a9b44eR9
so when this is evaluted:
```
HostStorage.includes(:reserved_rec)
```
it loads related model Reserve as is defined in ReservedMixin
https://github.com/ManageIQ/manageiq/blob/master/lib/reserved_mixin.rb#L4

and It causes loading Reserve model from app/models folder even if it is defined as stub in the migration.
cc @agrare 

Links
-----
case 2 needs https://github.com/ManageIQ/manageiq/pull/9827 

cc @Fryguy @jrafanie

Fryguy commented:
Thanks @lpichler .   My thoughts on these two:

1.  We probably should avoid loading MiqReport objects if possible.  That migration is notoriously bad, performance-wise, in particular with its YAML dump/load (see #9119), and I'm sure there are better ways to do it altogether.  So, I am fine with this exclusion.

2.  The Reserve model should probably live elsewhere...perhaps lib.  Its purpose is specifically for migrations.  So, I am fine with this exclusion.

@jrafanie Any other concerns on this (maybe with boot time stuff since it's in initializers)?

jrafanie commented:
> The Reserve model should probably live elsewhere...perhaps lib. Its purpose is specifically for migrations. So, I am fine with this exclusion.

I wonder if we should just move these files out of app and into lib.  ❓

Fryguy commented:
MiqExpression is not a model itself, so probably?  @imtayadeway   I'm already for moving both Reserve and ReserveMixin out of models and into lib.

imtayadeway commented:
@Fryguy Are you asking about moving MiqExpression too? I'm good with that....I'd like for it to be more like a library anyway :grin:

jrafanie commented:
@lpichler @Fryguy @imtayadeway So, yes, I think these should be moved but I'm ok with doing this PR temporarily until that is done so we don't have good_migrations complaining about it in the interim.

lpichler commented:
@Fryguy @jrafanie @imtayadeway I moved both files into lib folder, thanks!

imtayadeway commented:
@lpichler if you're moving MiqExpression, you might want to move the objects that live under it too in /app/models/miq_expression/. Also, if you're moving those, I wonder if we could fine a better home for the mixin that lives there too?

lpichler commented:
@imtayadeway which mixin do you have on mind ? filter_subst_mixin.rb ?

imtayadeway commented:
@lpichler that's the one!

lpichler commented:
> @lpichler tag. You now own all of MiqExpression 😆 /cc @jrafanie
Does it mean that I should fix all rubocop warnings :D ?

lpichler commented:
@imtayadeway @jrafanie @Fryguy 
I moved also files from app/models/miq_expression as @imtayadeway suggested. We also decided that filter_subst_mixin.rb will be on same place as it is now

Thanks!

jrafanie commented:
> 5 files checked, 204 offenses detected

ugh, moving legacy classes is rough.

We can fix these cops at a later time.

@lpichler can you update the PR description as it still references the first idea, the initializer?

lpichler commented:
@jrafanie updated, thanks

imtayadeway commented:
@lpichler @jrafanie would it make sense to move the corresponding specs in spec/models too?

jrafanie commented:
> @lpichler @jrafanie would it make sense to move the corresponding specs in spec/models too?

Yes!  Good point.

lpichler commented:
@imtayadeway yes, good idea.

@jrafanie  so I moved 
### MiqExpression
`spec/models/miq_expression_spec.rb` to `spec/lib/miq_expression_spec.rb`
`spec/models/miq_expression/field_spec.rb` to `spec/lib/miq_expression/field_spec.rb`
`spec/models/miq_expression/relative_datetime_spec.rb` to `spec/lib/miq_expression/relative_datetime_spec.rb`
no other related files exist.
### Reverse
there is no reserve_spec so I moved `reserved_mixin_spec.rb` to `spec/lib/reserved_mixin_spec.rb` as we moved reserved_mixin in https://github.com/ManageIQ/manageiq/pull/7348

thanks!

jrafanie commented:
@lpichler looks good, rebase, and let us know what you want to do about this one and #9827.  I don't understand why 9827 passed tests either.

lpichler commented:
@jrafanie rebased, I created https://github.com/ManageIQ/manageiq/pull/10050 for discussion why it is not failing

chrisarcand commented:
Excellent. The more non-model classes moved out of `/models` to `/lib` the happier I am 😄

blomquisg commented:
@lpichler This PR needs to be rebased.  Also, it has two other PRs depending on it (https://github.com/ManageIQ/manageiq/pull/9827 and https://github.com/ManageIQ/manageiq/pull/9688).

Any chance this can get some attention?  @jrafanie @Fryguy @gtanzillo 

/cc @agrare

lpichler commented:
@blomquisg rebased.

@jrafanie I will continue to investigate issue with good-migration gem in separate PR ( 'why it is not failing on CI')

so I think that this can be merged.

gtanzillo commented:
👍 LGTM
